lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,307 @@
+          unsigned cumulativeUpperBound = 1;
+          for (int k = j; k < tileOffsets[oldArgIndex + 1]; k++)
+            cumulativeUpperBound *= op.getStaticUpperBound()[j];
+          Value upperBound = rewriter.create<arith::ConstantIndexOp>(
+              op.getLoc(), cumulativeUpperBound);
+          mul = rewriter.create<arith::MulIOp>(op.getLoc(), b.getIndexType(),
+                                               add, upperBound);
+          add = rewriter.create<arith::AddIOp>(
+              op.getLoc(), b.getIndexType(), mul, op.getBody()->getArgument(j));
+        }
+        oper->setOperand(operandIndex, add);
+      }
+      operandIndex++;
+    }
+  }
+
+  for (int i = 1; i <= 3; i++) {
+    auto definingOp = brgemmOp.getOperand(i).getDefiningOp();
+    if (isa<memref::SubViewOp>(definingOp)) {
+      Value operand = definingOp->getOperand(0);
+
+      auto operandType = operand.getType();
+      auto resultType = dyn_cast<MemRefType>(brgemmOp.getOperand(i).getType());
+      auto reassociationIndex =
+          getReassociationIndices(originalShapes[i - 1], tilingVectors[i - 1]);
+
+      SmallVector<int64_t> shape;
+      for (size_t j = 0; j < tilingVectors[i - 1].size(); j++)
+        shape.append(tilingVectors[i - 1][j].begin(),
+                     tilingVectors[i - 1][j].end());
+      shape.append(
+          std::next(originalShapes[i - 1].begin(), tilingVectors[i - 1].size()),
+          originalShapes[i - 1].end());
+      int inductionVarBase = i - 1;
+      if (definingOp == brgemmOp.getOperand(3).getDefiningOp())
+        inductionVarBase = 0;
+      insertSubview(shape, operandType, resultType, reassociationIndex, operand,
+                    op, rewriter, definingOp, inductionVarBase);
+    }
+  }
+
+  return success();
+}
+
+bool getInnermostForLoops(Operation *rootOp,
+                          SmallVectorImpl<scf::ForallOp> &result) {
+  assert(rootOp != nullptr && "Root operation must not be a nullptr.");
+  bool rootEnclosesForAllloops = false;
+  for (Region &region : rootOp->getRegions()) {
+    for (Block &block : region.getBlocks()) {
+      for (Operation &op : block) {
+        bool enclosesPloops = getInnermostForLoops(&op, result);
+        rootEnclosesForAllloops |= enclosesPloops;
+        if (auto ploop = dyn_cast<scf::ForallOp>(op)) {
+          rootEnclosesForAllloops = true;
+
+          // Collect forall loop if it is an innermost one.
+          if (!enclosesPloops)
+            result.push_back(ploop);
+        }
+      }
+    }
+  }
+  return rootEnclosesForAllloops;
+}
+
+struct LoopInsertionPass
+    : public tpp::impl::LoopInsertionPassBase<LoopInsertionPass> {
+
+  using LoopInsertionPassBase::LoopInsertionPassBase;
+
+  void runOnOperation() override {
+    auto *parentOp = getOperation();
+    SmallVector<ForallOp> innermostForAllloops;
+    getInnermostForLoops(parentOp, innermostForAllloops);
+    for (ForallOp loop : innermostForAllloops)
```

Comment:
Why isn't this one a walk like the others?

Reply:
We need to parallelize the innermost forall loop, which is hard to detect with a walk. The upstream parallelization pass also does something similar, look for the innermost loop and invoke tiling on it.